### ISSUE TITLE
Support github repository trees

### ISSUE THREAD

**Issue by Jarred-Sumner** [Canop/broot#105]
**Did a discussion occur before?**
I posted in the chat room.

**Is your feature request related to a problem? Please describe.**
Navigating the file tree in GitHub repositories is slow (even with the `T` command) and often I would like to open or a copy a file into my text editor without cloning the whole repository.  

This is 100% _not_ `broot`'s problem, however `broot` would be an excellent tool to fix this, and I'm not the only person to bring this up: https://news.ycombinator.com/item?id=22005333

**Describe the solution you'd like**
```bash
broot https://github.com/Canop/broot
```
This would open the directory tree of the GitHub repo in `broot`. The functionality would be the same as the local filesystem, but opening the file would fetch the contents of the specific file & save it to a temporary directory(?)

I would use `broot` a lot more if it had this.

**Describe alternatives you've considered**
- Using "T" in Github. It's just not as good as `broot`.
- [OctoTree](https://chrome.google.com/webstore/detail/octotree/bkhaagjahfmjljalopjnoealnfndnagc?hl=en-US)

**Additional context**
I'm not really sure how one would implement this. Ideally, it would work without cloning the git repo, or having a dependency on `git`.

Without context on this codebase, I see two main issues:
- Reading the contents of a remote file. This answer looks helpful: https://stackoverflow.com/questions/2466735/how-to-sparsely-checkout-only-one-single-file-from-a-git-repository#answer-14779969
- Reading the metadata of the directory tree. For that, one could take inspiration from what [Octotree does](https://github.com/ovity/octotree/blob/6a362bcc4be41b6f2f93093ce17291b896d04527/src/adapters/github.js#L206). This would require auth'ing with the GitHub API however, which is a little bit of friction and a minor security concern (where do you store the access token? should it support private repositories too?)

**Comment by Canop** [comment]
This is interesting. I'll look at this as soon as possible.

**Comment by briangordon** [comment]
@Jarred-Sumner IMHO this problem should be solved at the filesystem level. Take a look at [GVFS](https://vfsforgit.org/). broot doesn't need to know anything about git because as far as it knows it's just browsing a regular directory and opening regular files. GVFS takes care of checking out files on the fly.

**Comment by Jarred-Sumner** [comment]
@briangordon I agree that from a purely technical perspective, a virtual Git filesystem is a good solution. From a product or requirements perspective though, it doesn't make sense.

The ephemerality & speed is important – I want to take a little code snippet from a repo somewhere and paste/edit a modified version of it into my codebase. Hunting around for the specific file & getting it into my text editor is too slow. This is a common workflow.

Cloning a repo or using GVFS is too much work. I don't want to keep the files on my computer anyway because then I have to think about where it would go (not a disk space issue). For this to change, GVFS would need to make every repository on GitHub browse-able in the filesystem so that I could `cd` into `/github/Canop/broot` or something (like a massive Dropbox folder). Also, GVFS has poor support for macOS and is really only designed for Big Company git repositories – not ephemeral browsing & searching.

**Comment by Canop** [comment]
Broot only shows real files, and I've decided to keep it that way: no virtual trees, be they archives or remote repositories (see https://dystroy.org/blog/why-broot-doesnt-enter-zip/). A shallow clone, or a virtual git filesystem as suggested above, gives you a real directory broot can browse. Closing.
